### DISCUSSION TITLE
'nVars' argument to addVarGroup must be greater than  or equal to 1. The bad DV is shapey

### DISCUSSION THREAD

**Discussion by Janfoam** [mdolab/dafoam#92] [General]
Hi, 

I get the error below after making an adjustment to the indexlist in the runscript.py. Can someone please say what I could be doing wrong?

Thanks

![image](https://user-images.githubusercontent.com/76228791/109841687-95a4e280-7c49-11eb-88f3-6f66e65a0a18.png)

**Comment by friedenhe** [comment]
What did you change?

**Reply by Janfoam** [reply]
Hi friendenhe, 

Thanks for your response. Apologies for the delay to respond.

I changed the indexlist in the runScript.py. I want to hold some of the points fixed instead of having all of them move freely with the pts[:, :, :] command. So for example, I would like to hold the trailing edge fixed. I am using the command on the DAFoam website indexList = pts[1:2, 3, 5:6].flatten(), but I get the message below. I have tried other combinations, but each time I get an error message. Looking online to better understand the Python syntax, I get that the format is a type of index splicing. However, I don't understand what I am doing wrong.  

Also, another question: the FFD propagates the additional points once the borders are provided. But how do you manipulate these additional points. Right now it seems its only possible to manipulate the borders? 

Apologies to bring this matter up again, but I had to step away from FFD-related content for a while to attend to some other matters. 

![image](https://user-images.githubusercontent.com/76228791/110221887-07419280-7ecf-11eb-8942-da366688db3d.png)

**Reply by friedenhe** [reply]
Your index is out of range, as in the error message. Python index starts with 0. I am not quite sure what you meant in the 2nd question.

**Reply by Janfoam** [reply]
Thank you Friedenhe, 

What I mean is this: We provide some of the points and the FFD automatically fills in the rest. We only have control over the points we provide through the FFD section of the genFFD.py script. The additional points that are automatically added we do not have control over those points to fix or move, because those are filled in automatically by the FFD. Please see screenshot below as to what I mean. Therefore, how does one fix or move the automatically generated points seeing those only appear after running python genFFD.py?

I have been trying to understand the indexing these last days, but still having roadblocks. Taking the block in the genFFD.py as an example: these are 8 points [0,1,2,3,4,5,6,7]. Points 4,5,6,7 are all related to the trailing edge whereas points 0,1,2,3 are all related to the leading edge. Based on what I understand from the numpy indexing literature: https://numpy.org/doc/stable/reference/arrays.indexing.html, the basic slice syntax is i:j:k where i is the starting index, j is the stopping index, and k is the step (k is not equal to zero).  

For example, to hold the trailing edge fixed [4:7:1] which means start at four, stop at 7 with increments of 1. However, I get an error. Can you please clarify with some details what I am missing/failing to consider?

![image](https://user-images.githubusercontent.com/76228791/110551190-61439180-8135-11eb-8b23-e6511cc852f7.png)
![image](https://user-images.githubusercontent.com/76228791/110551222-6c96bd00-8135-11eb-897f-a64b95baa74d.png)

**Reply by friedenhe** [reply]
Your understanding of how genFFD.py works is not quite right. You prescribe CORNER points, and the genFFD.py will generate the rest. Once you load the FFD points in runScript.py and provide it to DVGeo, you have FULL control of all FFD points defined in wingFFD.xyz. In the case you showed, you can control 5x5x2=50 FFD points, instead of the 8 corner points. Also, if you do 4:7:1 the index will go from 4 to 6, instead of 4 to 7.
